File: Medium/2026-03-23-1594-Maximum-Non-Negative-Product-in-a-Matrix/solution.java
class Solution {
    private static final int MOD = 1000000007;

    public int maxProductPath(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        long[][] maxProd = new long[m][n];
        long[][] minProd = new long[m][n];

        maxProd[0][0] = grid[0][0];
        minProd[0][0] = grid[0][0];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 && j == 0) continue;

                long maxVal = Long.MIN_VALUE;
                long minVal = Long.MAX_VALUE;

                if (i > 0) {
                    maxVal = Math.max(maxVal, maxProd[i - 1][j] * grid[i][j]);
                    minVal = Math.min(minVal, minProd[i - 1][j] * grid[i][j]);
                }
                if (j > 0) {
                    maxVal = Math.max(maxVal, maxProd[i][j - 1] * grid[i][j]);
                    minVal = Math.min(minVal, minProd[i][j - 1] * grid[i][j]);
                }

                maxProd[i][j] = maxVal;
                minProd[i][j] = minVal;
            }
        }

        long result = maxProd[m - 1][n - 1];
        return result < 0 ? -1 : (int) (result % MOD);
    }
}

File: Medium/2026-03-23-1594-Maximum-Non-Negative-Product-in-a-Matrix/solution.py
class Solution:
    def maxProductPath(self, grid: List[List[int]]) -> int:
        MOD = 10**9 + 7
        m, n = len(grid), len(grid[0])
        
        dp = [[(0, 0) for _ in range(n)] for _ in range(m)]
        dp[0][0] = (grid[0], grid[0])
        
        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0:
                    continue
                max_product = float('-inf')
                min_product = float('inf')
                
                if i > 0:
                    max_product = max(max_product, dp[i-1][j][0] * grid[i][j])
                    min_product = min(min_product, dp[i-1][j][1] * grid[i][j])
                if j > 0:
                    max_product = max(max_product, dp[i][j-1][0] * grid[i][j])
                    min_product = min(min_product, dp[i][j-1][1] * grid[i][j])
                
                dp[i][j] = (max_product, min_product)
        
        max_product = dp[m-1][n-1][0]
        return max_product % MOD if max_product >= 0 else -1